### PR TITLE
[dy] Update pipeline execution

### PR DESCRIPTION
**Pull request by dy46**
# Summary

Two main features
* Allow cancellation of pipeline. 
* Show blocks being run in the UI during pipeline execution

I also updated how we show the execution output. Instead of just logging print statements to the pipeline execution area, the output in the code block will update once the block finished running. 

# Tests

tested locally.

pipeline_execution

execution_cancel

cc: @wangxiaoyou1993

### CODE REVIEW

**Comment by wangxiaoyou1993** [comment]
Could you attach screenshots if there're UI changes?

**Comment by dy46** [comment]
@wangxiaoyou1993 sorry for the delay, but i added some gifs

File: mage_ai/server/websocket.py [hunk 1]
```diff
@@ -109,16 +124,40 @@ def publish_message(
 
                 WebSocketServer.running_executions_mapping[msg_id] = value
             else:
+                # TODO: save config for other kernel types.
+                def save_pipeline_config() -> str:
```
Comment (wangxiaoyou1993) [inline mage_ai/server/websocket.py:128]: this `on_message` method is too long. could you refactor it?

File: mage_ai/server/websocket.py [hunk 2]
```diff
@@ -109,16 +124,40 @@ def publish_message(
...
                 def run_pipeline() -> None:
                     try:
-                        asyncio.run(pipeline.execute(log_func=publish_message, redirect_outputs=True))
-                        publish_message(f'Pipeline {pipeline.uuid} execution complete.', 'idle')
+                        pipeline.execute_sync(
+                            log_func=publish_message,
+                        )
+                        publish_message(
+                            f'Pipeline {pipeline.uuid} execution complete.\n'
+                            'You can see the code block output in the corresponding code block.',
```
Comment (wangxiaoyou1993) [inline mage_ai/server/websocket.py:147]: let's specify the keyword argument names.

File: mage_ai/server/websocket.py [hunk 3]
```diff
@@ -109,16 +124,40 @@ def publish_message(
...
+                        )
                     except Exception:
                         trace = traceback.format_exc().splitlines()
                         publish_message(f'Pipeline {pipeline.uuid} execution failed with error:')
                         publish_message(trace, 'idle')
 
-                threading.Thread(target=run_pipeline).start()
+                    delete_pipeline_copy_config(config_copy_path)
+
+                save_pipeline_config()
```
Comment (wangxiaoyou1993) [inline mage_ai/server/websocket.py:157]: Could you add comments for why we need to save pipeline config?

File: mage_ai/data_preparation/models/block.py [hunk 4]
```diff
@@ -34,26 +33,39 @@ async def run_blocks(
...
-            block.run_tests(update_tests=False)
     tasks = dict()
     blocks = Queue()
 
+    def create_block_task(block: 'Block'):
+        async def execute_and_run_tests():
+            prefix = f'[{block.uuid}]'
+            try:
+                if log_func:
+                    log_func(f'{prefix} Executing {block.type} block...', execution_state='busy', block_uuid=block.uuid)
```
Comment (wangxiaoyou1993) [inline mage_ai/data_preparation/models/block.py:51]: nit: break the line if the line is too long.

File: mage_ai/data_preparation/models/block.py [hunk 5]
```diff
@@ -113,17 +126,26 @@ def run_blocks_sync(
...
             continue
-        block.execute_sync(
-            analyze_outputs=analyze_outputs,
-            redirect_outputs=redirect_outputs,
-            run_all_blocks=True,
-        )
+        try:
+            prefix = f'[{block.uuid}]'
+            if log_func:
+                log_func(f'{prefix} Executing {block.type} block...', execution_state='busy', block_uuid=block.uuid)
```
Comment (wangxiaoyou1993) [inline mage_ai/data_preparation/models/block.py:132]: nit: same. break the line

File: mage_ai/server/execution_manager.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
+from distutils.dir_util import copy_tree
+import os
+import shutil
+
+class PipelineExecution():
```
Comment (wangxiaoyou1993) [inline mage_ai/server/execution_manager.py:5]: Could you add some comments to this class since it's not easy to understand what it does without context?

**Comment by dy46** [comment]
@wangxiaoyou1993 addressed the comments

File: mage_ai/server/execution_manager.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
+from distutils.dir_util import copy_tree
+from mage_ai.data_preparation.models.pipeline import Pipeline
+from typing import Callable
+import multiprocessing
+import os
+import shutil
+
+'''
```
Comment (wangxiaoyou1993) [inline mage_ai/server/execution_manager.py:8]: Let's put the comments below the class name and use double quotes to wrap the comments.

File: mage_ai/server/execution_manager.py (hunk 6 above)
Comment (wangxiaoyou1993) [reply mage_ai/server/execution_manager.py:14]: You can remove the `()` after class name if it doesn't have parent classes.

File: mage_ai/server/execution_manager.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+class PipelineExecution():
+    def __init__(self):
+        self.current_pipeline_process: multiprocessing.Process = None
+        self.previous_config_path: str = None
+
+pipeline_execution = PipelineExecution()
+
+
+def set_current_pipeline_process(process: multiprocessing.Process) -> None:
+    '''
```
Comment (wangxiaoyou1993) [inline mage_ai/server/execution_manager.py:23]: Same here. Let's use double quotes.

File: mage_ai/server/websocket.py [hunk 9]
```diff
@@ -213,3 +141,148 @@ def send_message(self, message: dict) -> None:
...
+                delete_pipeline_copy_config(config_copy_path)
+
+            reset_execution_manager()
+
+            # The pipeline state can potentially break when the execution is cancelled,
+            # so we save the pipeline config before execution if the user cancels the excecution.
+            save_pipeline_config()
+            proc = multiprocessing.Process(target=run_pipeline)
+            proc.start()
+            set_current_pipeline_process(proc)
```
Comment (wangxiaoyou1993) [inline mage_ai/server/websocket.py:288]: nit: new line

**Review by wangxiaoyou1993 — APPROVED** [review]

**Pull request merged**